### PR TITLE
fix(rome_js_analyze): false positive useShorthandArrayTypes

### LINKED ISSUE #3111: 🐛 ts/useShorthandArrayType is not taking into consideration the need for parenthesis
### Environment information

Running the analyzer from the  ```main``` branch.

### What happened?

The lint rule ts/useShorthandArrayType is not taking into consideration the need for parenthesis when creating the suggested fix.

Example:

```js
let v = [] as Array<(...args: E) => void>;
console.log(v);
```

https://play.rome.tools/?code=let+v+%3D+%5B%5D+as+Array%3C%28...args%3A+E%29+%3D%3E+void%3E%3B%0Aconsole.log%28v%29%3B&lineWidth=80&indentStyle=tab&quoteStyle=double&indentWidth=2&sourceType=module&typescript=true&jsx=false#bABlAHQAIAB2ACAAPQAgAFsAXQAgAGEAcwAgAEEAcgByAGEAeQA8ACgALgAuAC4AYQByAGcAcwA6ACAARQApACAAPQA+ACAAdgBvAGkAZAA+ADsACgBjAG8AbgBzAG8AbABlAC4AbABvAGcAKAB2ACkAOwA=

Suggested fix here is:

```
error[[ts/useShorthandArrayType](https://rome.tools/docs/lint/rules/useShorthandArrayType/)]: Use shorthand T[] syntax instead of Array<T> syntax.
  ┌─ main.ts:1:15
  │
1 │ let v = [] as Array<(...args: E) => void>;
  │               ^^^^^^^^^^^^^^^^^^^^^^^^^^^

Suggested fix: Use shorthand T[] syntax to replace
    | @@ -1,2 +1,2 @@
0   | - let v = [] as Array<(...args: E) => void>;
  0 | + let v = [] as (...args: E) => void[];
1 1 |   console.log(v);
```

### Expected result

I was expecting the suggested fix to be 

```
let v = [] as ((...args: E) => void)[]; // <- see wrapping parenthesis
```

### Code of Conduct

- [X] I agree to follow Rome's Code of Conduct

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
## Summary
1. Fixes #3111 

## Test Plan
1. Checking newly added snapshot

### CODE REVIEW

File: crates/rome_js_analyze/tests/specs/ts/useShorthandArrayType.ts.snap [hunk 1]
```diff
@@ -15,7 +16,12 @@ let invalid2: Promise<Array<string>>;
 let invalid3: Array<Foo<Bar>>;
 let invalid4: Array<[number, number]>;
 
-
+// valid
+let valid5: Array<string & number>;
```
Comment (MichaReiser) [inline crates/rome_js_analyze/tests/specs/ts/useShorthandArrayType.ts.snap:20]: Would you mind adding some valid and invalid tests for if the type is parenthesized.

File: crates/rome_js_analyze/src/analyzers/ts/use_shorthand_array_type.rs [hunk 2]
```diff
@@ -112,8 +112,14 @@ fn convert_to_array_type(type_arguments: TsTypeArguments) -> Option<TsType> {
...
                 let element_type = match &param {
-                    TsType::TsUnionType(_) => None,
-                    TsType::TsTypeOperatorType(_) => None,
+                    // Intersection or higher types
+                    TsType::TsUnionType(_)
+                    | TsType::TsIntersectionType(_)
+                    | TsType::TsFunctionType(_)
+                    | TsType::TsConstructorType(_)
+                    | TsType::TsConditionalType(_)
+                    | TsType::TsTypeOperatorType(_) => None,
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/ts/use_shorthand_array_type.rs:121]: What about mapped types and `infer`?

File: crates/rome_js_analyze/tests/specs/ts/useShorthandArrayType.ts.snap (hunk 1 above)
Comment (IWANABETHATGUY) [reply crates/rome_js_analyze/tests/specs/ts/useShorthandArrayType.ts.snap:20]: I guess all parenthesis typed should be invalid which means should convert to shorthandArrayTypes, 
https://play.rome.tools/?code=type+res+%3D+Array%3C%28%28a%29%29%3E&lineWidth=80&indentStyle=tab&quoteStyle=double&indentWidth=2&sourceType=module&typescript=true&jsx=false#dAB5AHAAZQAgAHIAZQBzACAAPQAgAEEAcgByAGEAeQA8ACgAKABhACkAKQA+AA==. I will add it later.

Comment (MichaReiser) [reply crates/rome_js_analyze/tests/specs/ts/useShorthandArrayType.ts.snap:20]: Rome's analyzer should work independently of Rome's formatter. Users can decide to only adopt one of the two.

Comment (IWANABETHATGUY) [reply crates/rome_js_analyze/tests/specs/ts/useShorthandArrayType.ts.snap:20]: So which behavior do you expect, e.g.
```js
let valid5: Array<(string & number)>;
```
1. keep as is if the inner type should not convert to `shorthandArrayType`
```js
let valid: Array<(foo | bar)>;
```

2. Convert it to shorthandArraytype. (this is current behavior same as rome classic)
```js
let valid: (foo | bar)[];
```

Comment (ematipico) [reply crates/rome_js_analyze/tests/specs/ts/useShorthandArrayType.ts.snap:20]: The old rule used to apply the following fix:

```ts
let valid5: (string & number)[];
```

File: crates/rome_js_analyze/src/analyzers/ts/use_shorthand_array_type.rs (hunk 2 above)
Comment (IWANABETHATGUY) [reply crates/rome_js_analyze/src/analyzers/ts/use_shorthand_array_type.rs]: Done

File: crates/rome_js_analyze/tests/specs/ts/useShorthandArrayType.ts.snap (hunk 1 above)
Comment (IWANABETHATGUY) [reply crates/rome_js_analyze/tests/specs/ts/useShorthandArrayType.ts.snap:20]: Done

**Comment by IWANABETHATGUY** [comment]
All the behavior has been aligned to rome classic

**Review by ematipico — APPROVED** [review]

**Review by leops — APPROVED** [review]

**Pull request merged**